homeassistant/components/sensor/qnap.py

```diff
@@ -0,0 +1,413 @@
+
+ATTR_DRIVE = 'Drive'
+ATTR_DRIVE_SIZE = 'Drive Size'
+ATTR_IP = 'IP Address'
+ATTR_MAC = 'MAC Address'
+ATTR_MASK = 'Mask'
+ATTR_MAX_SPEED = 'Max Speed'
+ATTR_MEMORY_SIZE = 'Memory Size'
+ATTR_MODEL = 'Model'
+ATTR_NAME = 'Name'
+ATTR_PACKETS_TX = 'Packets (TX)'
+ATTR_PACKETS_RX = 'Packets (RX)'
+ATTR_PACKETS_ERR = 'Packets (Err)'
+ATTR_SERIAL = 'Serial #'
+ATTR_TYPE = 'Type'
+ATTR_UPTIME = 'Uptime'
+ATTR_VOLUME_SIZE = 'Volume Size'
+
+CONF_DRIVES = 'drives'
+CONF_NICS = 'nics'
+CONF_VOLUMES = 'volumes'
+DEFAULT_NAME = 'QNAP'
+DEFAULT_PORT = 8080
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=1)
+
+_HEALTH_MON_COND = {
+    'status': ['Status', None, 'mdi:checkbox-marked-circle-outline'],
+}
+_CPU_MON_COND = {
+    'cpu_temp': ['CPU Temperature', TEMP_CELSIUS, 'mdi:thermometer'],
+    'cpu_usage': ['CPU Usage', '%', 'mdi:chip'],
+}
+_MEMORY_MON_COND = {
+    'memory_free': ['Memory Available', 'GB', 'mdi:memory'],
+    'memory_used': ['Memory Used', 'GB', 'mdi:memory'],
+    'memory_percent_used': ['Memory Usage', '%', 'mdi:memory'],
+}
+_NETWORK_MON_COND = {
+    'network_link_status': ['Network Link', None,
+                            'mdi:checkbox-marked-circle-outline'],
+    'network_tx': ['Network Up', 'MB/s', 'mdi:upload'],
+    'network_rx': ['Network Down', 'MB/s', 'mdi:download'],
+}
+_DRIVE_MON_COND = {
+    'drive_smart_status': ['SMART Status', None,
+                           'mdi:checkbox-marked-circle-outline'],
+    'drive_temp': ['Temperature', TEMP_CELSIUS, 'mdi:thermometer'],
+}
+_VOLUME_MON_COND = {
+    'volume_size_used': ['Used Space', 'GB', 'mdi:chart-pie'],
+    'volume_size_free': ['Free Space', 'GB', 'mdi:chart-pie'],
+    'volume_percentage_used': ['Volume Used', '%', 'mdi:chart-pie'],
+}
+
+_MONITORED_CONDITIONS = list(_HEALTH_MON_COND.keys()) + \
+                        list(_CPU_MON_COND.keys()) + \
+                        list(_MEMORY_MON_COND.keys()) + \
+                        list(_NETWORK_MON_COND.keys()) + \
+                        list(_DRIVE_MON_COND.keys()) + \
+                        list(_VOLUME_MON_COND.keys())
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PROTOCOL, default='http'): vol.In(['http', 'https']),
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_MONITORED_CONDITIONS):
+        vol.All(cv.ensure_list, [vol.In(_MONITORED_CONDITIONS)]),
+    vol.Optional(CONF_NICS, default=None): cv.ensure_list,
+    vol.Optional(CONF_DRIVES, default=None): cv.ensure_list,
+    vol.Optional(CONF_VOLUMES, default=None): cv.ensure_list,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the QNAP NAS sensor."""
+    # pylint: disable=too-many-locals
+    def run_setup(event):
+        """Wait until HASS is fully initialized before creating sensors."""
```

Comment:
Please remove this functionality and set it up inside `setup_platform`. All entities should be set up when HOMEASSISTANT_START fires.